internal/plugin/v1/provider.go

```diff
@@ -7,11 +7,47 @@ type ProviderOptions struct {
 	Name string
 }
 
+// ProviderResponse is the response from the provider for a given secret request
+type ProviderResponse struct {
+	Value []byte
+	Error error
+}
+
 // Provider is the interface used to obtain values from a secret vault backend.
 type Provider interface {
 	// GetName returns the name that the Provider was instantiated with
 	GetName() string
 
+	// GetValues takes in variable ids and returns their resolved values
+	GetValues(ids ...string) (map[string]ProviderResponse, error)
+}
+
+type singleValueProvider interface {
 	// GetValue takes in an id of a variable and returns its resolved value
 	GetValue(id string) ([]byte, error)
 }
+
+// GetValues takes in variable ids and returns their resolved values by making sequential
+// getValueCallArgs to a singleValueProvider.
+// This is a convenience function since most providers with batch retrieval capabilities
+// will have need the exact same code. Note: most internal providers simply use this
+// function in their implementation of the Provider interface's GetValues method.
+func GetValues(
+	p singleValueProvider,
+	ids ...string,
+) (map[string]ProviderResponse, error) {
+	responses := map[string]ProviderResponse{}
+
+	for _, id := range ids {
+		if _, ok := responses[id]; ok {
+			continue
+		}
+
+		var pr ProviderResponse
+		pr.Value, pr.Error = p.GetValue(id)
```

Comment:
Not a blocker but this would be nicer to see in the more-common `foo := ProviderResponse{}` pattern.